core/src/apps/cardano/helpers/tx_builder.py

```diff
@@ -0,0 +1,194 @@
+from micropython import const
+
+from trezor.crypto import hashlib
+from trezor.messages import CardanoAddressType
+from trezor.messages.CardanoAddressParametersType import CardanoAddressParametersType
+
+from apps.common import cbor
+
+from ..address import derive_address_bytes, get_address_bytes_unsafe
+from ..certificates import cborize_certificate
+from .lazy_collection import DICT, LIST, LazyCollection
+
+if False:
+    from trezor.messages.CardanoSignTx import CardanoSignTx
+    from trezor.messages.CardanoTxInput import CardanoTxInput
+    from trezor.messages.CardanoTxOutput import CardanoTxOutput
+    from trezor.messages.CardanoAssetGroup import CardanoAssetGroup
+    from trezor.messages.CardanoToken import CardanoToken
+    from trezor.messages.CardanoTxCertificate import CardanoTxCertificate
+    from trezor.messages.CardanoTxWithdrawal import CardanoTxWithdrawal
+    from .. import seed
+    from typing import Any, Optional
+
+INPUT = const(0)
+OUTPUT = const(1)
+ASSET_GROUP = const(2)
+TOKEN = const(3)
+CERTIFICATE = const(4)
+WITHDRAWAL = const(5)
+
+
+class TxBuilder:
+    def __init__(self, keychain: seed.Keychain, msg: CardanoSignTx):
+        tx_body_map_item_count = 3 + sum(
+            1
+            for entry in [
+                msg.ttl,
+                msg.certificates_count,
+                msg.withdrawals_count,
+                msg.has_auxiliary_data,
+                msg.validity_interval_start,
+            ]
+            if bool(entry)
+        )
+        self.state = [LazyCollection(DICT, tx_body_map_item_count)]
+        self.serializer = cbor.encode_streamed(self.state[0])
+        self.hashfn = hashlib.blake2b(outlen=32)
+        self._update_tx_hash()
+        self.keychain = keychain
+        self.msg = msg
+        self.expected_item_kind: Optional[int] = None
+
+    def _update_tx_hash(self) -> None:
+        while (chunk := next(self.serializer)) is not None:
+            self.hashfn.update(chunk)
+
+    def get_tx_hash(self) -> bytes:
+        assert self.state[0].is_filled(), "The hash calculation is not finished yet"
+        return self.hashfn.digest()
+
+    def _start_collection_at_key(
+        self,
+        key: Any,
+        collection: LazyCollection,
+        next_kind: int,
```

Comment:
I'd consider using an enum for the next_kind and current_kind type - that way also the function signature could be more explicit, rather than falling back to the generic "int" type

Reply:
I'm afraid MicroPython doesn't support enums out of the box (getting `ImportError: no module named 'enum'`). Maybe that's the reason of all the `const`s in `cbor.py`.
I can introduce a type alias for `int` though if you think that helps.

Reply:
I see, seems sensible to at least alias the int type then